### PR TITLE
Dequeue as a flux

### PR DESCRIPTION
**Pull request by mbaechler**
For review only, not to be merged

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[2ee56ece8485af8339e23d4b2b052a7581fa6189] [ERROR] Errors: 
[2ee56ece8485af8339e23d4b2b052a7581fa6189] [ERROR]   CassandraRabbitMQJamesServerTest$MailsShouldBeWellReceived.mailsShouldBeWellReceived » IllegalState

java.lang.IllegalStateException: Queue is empty: Expected one element to be available from the Reactive Streams source
```

**Comment by mbaechler** [comment]
screenshot from 2018-12-06 09-12-58

3 hours run with this branch

File: backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitChannelPoolImpl.java [hunk 1]
```diff
@@ -89,6 +91,11 @@ public RabbitChannelPoolImpl(RabbitMQConnectionFactory factory) {
         pool = new GenericObjectPool<>(pooledChannelsFactory);
     }
 
+    @Override
+    public Flux<AcknowledgableDelivery> receive(String queueName) {
+        return null;
```
Comment (chibenwa) [inline backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitChannelPoolImpl.java]: Why null?

Comment (chibenwa) [reply backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitChannelPoolImpl.java]: Maybe we could explicify this?

Comment (mbaechler) [reply backends-common/rabbitmq/src/main/java/org/apache/james/backend/rabbitmq/RabbitChannelPoolImpl.java]: To be honest, I will rework it for many reasons but it's not for today

**Comment by rouazana** [comment]
it has been released! https://search.maven.org/search?q=reactor-rabbitmq

**Comment by chibenwa** [comment]
```
[ff35e2122dbb51da40dfa93c14b02982e543481b] java.lang.IllegalStateException: Queue is empty: Expected one element to be available from the Reactive Streams source.

[ff35e2122dbb51da40dfa93c14b02982e543481b] [ERROR]   CassandraRabbitMQJamesServerTest$MailsShouldBeWellReceived.mailsShouldBeWellReceived » IllegalState
```

**Comment by rouazana** [comment]
test this please

**Comment by mbaechler** [comment]
```
[f484a572e1c43d1ca408d38c5b18a9412868cc25] [ERROR] Tests run: 9, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 27.307 s <<< FAILURE! - in org.apache.james.jmap.cassandra.access.CassandraAccessTokenRepositoryTest
[f484a572e1c43d1ca408d38c5b18a9412868cc25] [ERROR] validTokenMustBeRetrieved  Time elapsed: 0.071 s  <<< ERROR!
[f484a572e1c43d1ca408d38c5b18a9412868cc25] java.util.concurrent.CompletionException: org.apache.james.jmap.api.access.exceptions.InvalidAccessToken: c627480a-ceb5-4278-9e4c-498d9352aa40 is missing
[f484a572e1c43d1ca408d38c5b18a9412868cc25] Caused by: org.apache.james.jmap.api.access.exceptions.InvalidAccessToken: c627480a-ceb5-4278-9e4c-498d9352aa40 is missing
[f484a572e1c43d1ca408d38c5b18a9412868cc25] 
```

test this please

**Comment by mbaechler** [comment]
```

[ed94ba479531222b42cdd54ad87e161775909952] [ERROR] shouldNotRejectContentWithStartTls  Time elapsed: 53.996 s  <<< ERROR!
[ed94ba479531222b42cdd54ad87e161775909952] java.lang.IllegalStateException: Queue is empty: Expected one element to be available from the Reactive Streams source.
[ed94ba479531222b42cdd54ad87e161775909952] 
```

**Comment by trantienduchn** [comment]
```
[03e87b71b67e35451e6fa2c3c8c2fd365fffc888] [ERROR] Errors: 
[03e87b71b67e35451e6fa2c3c8c2fd365fffc888] [ERROR]   SpamAssassinContract.spamShouldBeDeliveredInSpamMailboxWhenSameMessageHasAlreadyBeenMovedToSpam:146 » ConditionTimeout
[03e87b71b67e35451e6fa2c3c8c2fd365fffc888] [INFO] 
[03e87b71b67e35451e6fa2c3c8c2fd365fffc888] [ERROR] Tests run: 655, Failures: 0, Errors: 1, Skipped: 2
```
test this please

**Comment by mbaechler** [comment]
```
[e07a356e108019e283a0a29c4fb3f616bd07bc11] [ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 64.401 s <<< FAILURE! - in org.apache.james.mpt.smtp.RabbitMQForwardSmtpTest
[e07a356e108019e283a0a29c4fb3f616bd07bc11] [ERROR] forwardingAnEmailShouldWork  Time elapsed: 62.162 s  <<< ERROR!
[e07a356e108019e283a0a29c4fb3f616bd07bc11] org.awaitility.core.ConditionTimeoutException: 
[e07a356e108019e283a0a29c4fb3f616bd07bc11] Assertion condition defined as a org.apache.james.mpt.smtp.ForwardSmtpTest 1 expectation failed.
[e07a356e108019e283a0a29c4fb3f616bd07bc11] JSON path [0].from doesn't match.
[e07a356e108019e283a0a29c4fb3f616bd07bc11] Expected: matthieu@yopmail.com
[e07a356e108019e283a0a29c4fb3f616bd07bc11]   Actual: null
[e07a356e108019e283a0a29c4fb3f616bd07bc11]  within 60 seconds.
[e07a356e108019e283a0a29c4fb3f616bd07bc11] 	at org.apache.james.mpt.smtp.ForwardSmtpTest.forwardingAnEmailShouldWork(ForwardSmtpTest.java:78)
[e07a356e108019e283a0a29c4fb3f616bd07bc11] Caused by: java.lang.AssertionError: 
[e07a356e108019e283a0a29c4fb3f616bd07bc11] 1 expectation failed.
[e07a356e108019e283a0a29c4fb3f616bd07bc11] JSON path [0].from doesn't match.
[e07a356e108019e283a0a29c4fb3f616bd07bc11] Expected: matthieu@yopmail.com
[e07a356e108019e283a0a29c4fb3f616bd07bc11]   Actual: null
[e07a356e108019e283a0a29c4fb3f616bd07bc11] 
[e07a356e108019e283a0a29c4fb3f616bd07bc11] 	at org.apache.james.mpt.smtp.ForwardSmtpTest.lambda$null$0(ForwardSmtpTest.java:80)
[e07a356e108019e283a0a29c4fb3f616bd07bc11] 	at org.apache.james.mpt.smtp.ForwardSmtpTest.lambda$forwardingAnEmailShouldWork$1(ForwardSmtpTest.java:79)
[e07a356e108019e283a0a29c4fb3f616bd07bc11] 
[e07a356e108019e283a0a29c4fb3f616bd07bc11] [INFO] 
[e07a356e108019e283a0a29c4fb3f616bd07bc11] [INFO] Results:
[e07a356e108019e283a0a29c4fb3f616bd07bc11] [INFO] 
[e07a356e108019e283a0a29c4fb3f616bd07bc11] [ERROR] Errors: 
[e07a356e108019e283a0a29c4fb3f616bd07bc11] [ERROR]   ForwardSmtpTest.forwardingAnEmailShouldWork:78 » ConditionTimeout Assertion co...
```

test this please

**Comment by mbaechler** [comment]
test this please

**Comment by mbaechler** [comment]
```

[98d07cdf89f7ae07dfe554379a49d71f1a1ca19e] [INFO] 
[98d07cdf89f7ae07dfe554379a49d71f1a1ca19e] [ERROR] Errors: 
[98d07cdf89f7ae07dfe554379a49d71f1a1ca19e] [ERROR]   SendMDNMethodTest.sendMDNShouldSendAMDNBackToTheOriginalMessageAuthor:311 » ConditionTimeout
[98d07cdf89f7ae07dfe554379a49d71f1a1ca19e] [INFO] 
```

Tried the whole module testsuite on my machine, 1 failure, 2 success ...

test this please

**Comment by mbaechler** [comment]
```

[38455f043a720dde153ccb0b2c7d470eefeb7ead] [ERROR]   QuotaMailingTest.shouldSendANoticeWhenThresholdExceeded:114 » ConditionTimeout
[38455f043a720dde153ccb0b2c7d470eefeb7ead] [ERROR]   VacationIntegrationTest.jmapVacationShouldGenerateAReplyWhenActive:131 » ConditionTimeout
```

what the hell is wrong with that PR ?!

test this please

**Comment by chibenwa** [comment]
test this please

**Comment by mbaechler** [comment]
```
[38455f043a720dde153ccb0b2c7d470eefeb7ead] [ERROR] sendMDNShouldSendAMDNBackToTheOriginalMessageAuthor  Time elapsed: 124.778 s  <<< ERROR!
[38455f043a720dde153ccb0b2c7d470eefeb7ead] org.awaitility.core.ConditionTimeoutException: Condition with lambda expression in org.apache.james.jmap.methods.integration.SendMDNMethodTest was not fulfilled within 120 seconds.
[38455f043a720dde153ccb0b2c7d470eefeb7ead] 	at org.apache.james.jmap.methods.integration.SendMDNMethodTest.sendMDNShouldSendAMDNBackToTheOriginalMessageAuthor(SendMDNMethodTest.java:310)
[38455f043a720dde153ccb0b2c7d470eefeb7ead] 
[38455f043a720dde153ccb0b2c7d470eefeb7ead] [INFO] 
[38455f043a720dde153ccb0b2c7d470eefeb7ead] [INFO] Results:
[38455f043a720dde153ccb0b2c7d470eefeb7ead] [INFO] 
[38455f043a720dde153ccb0b2c7d470eefeb7ead] [ERROR] Errors: 
[38455f043a720dde153ccb0b2c7d470eefeb7ead] [ERROR]   SendMDNMethodTest.sendMDNShouldSendAMDNBackToTheOriginalMessageAuthor:310 » ConditionTimeout
```

**Comment by mbaechler** [comment]
it's green but it's probably an accident, let's get 3 green build in a row

test this please

**Comment by mbaechler** [comment]
```
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] java.lang.OutOfMemoryError: Java heap space
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] 	at java.util.Arrays.copyOf(Arrays.java:3332)
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] 	at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:124)
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] 	at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:596)
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] 	at java.lang.StringBuffer.append(StringBuffer.java:367)
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] 	at java.io.BufferedReader.readLine(BufferedReader.java:370)
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] 	at java.io.BufferedReader.readLine(BufferedReader.java:389)
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] 	at org.assertj.core.internal.Diff.linesFromBufferedReader(Diff.java:97)
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] 	at org.assertj.core.internal.Diff.diff(Diff.java:85)
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] 	at org.assertj.core.internal.Diff.diff(Diff.java:51)
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] 	at org.assertj.core.internal.InputStreams.assertSameContentAs(InputStreams.java:68)
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] 	at org.assertj.core.api.AbstractInputStreamAssert.hasSameContentAs(AbstractInputStreamAssert.java:78)
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] 	at org.apache.james.blob.cassandra.CassandraBlobsDAOTest.blobStoreShouldSupport100MBBlob(CassandraBlobsDAOTest.java:86)
```

completely unrelated

test this please

**Comment by chibenwa** [comment]
test this please (was :red_circle: )

`[ERROR]   QuotaMailingTest.shouldSendANoticeWhenThresholdExceeded:114 » ConditionTimeout`

`[INFO] Apache James :: Server :: JMAP :: RabbitMQ + Swift + Cassandra Integration testing FAILURE [06:40 min]`

**Comment by chibenwa** [comment]
test this please ( :green_apple:  )

**Comment by chibenwa** [comment]
test this please ( :green_apple:  )

**Comment by chibenwa** [comment]
```
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] [ERROR] Errors: 
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] [ERROR]   QuotaMailingTest.shouldSendANoticeWhenThresholdExceeded:114 » ConditionTimeout
```

test this please

**Comment by chibenwa** [comment]
```
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] [ERROR]   MailQueueContract.concurrentEnqueueDequeueWithAckNackShouldNotFail:430 
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] Expected size:<50> but was:<49> in:
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7] <["name0-0",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name9-0",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name1-0",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name7-0",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name3-0",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name8-0",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name6-0",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name2-0",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name5-0",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name4-0",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name3-3",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name4-3",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name1-3",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name6-3",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name8-3",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name9-3",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name5-3",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name2-3",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name7-3",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name0-3",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name0-6",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name4-6",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name3-6",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name2-6",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name1-6",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name6-6",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name5-6",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name8-6",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name9-6",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name7-6",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name1-9",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name7-9",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name2-9",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name8-9",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name6-9",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name4-12",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name5-12",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name5-9",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name7-12",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name9-9",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name4-9",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name3-9",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name0-12",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name9-12",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name0-9",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name8-12",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name2-12",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name1-12",
[8fcf6e422a2cc72d27952976f8ae1ab94bea0ba7]     "name6-12"]>
```

test this please

Was :red_circle:

**Comment by chibenwa** [comment]
test this please

wa :green_apple:

**Comment by chibenwa** [comment]
test this please :red_circle: 

` QuotaMailingTest.shouldSendANoticeWhenThresholdExceeded:114 » ConditionTimeout`

**Comment by mbaechler** [comment]
This fix could help : https://github.com/reactor/reactor-rabbitmq/pull/56

**Comment by mbaechler** [comment]
Rebased on master, let's see if it's green now

**Comment by mbaechler** [comment]
green, we'll try to get 3 in a row

test this please

**Comment by mbaechler** [comment]
ok, green twice, let's try a third time

test this please

**Comment by mbaechler** [comment]
```

[f77ff50d2bcf5e2a9421364f8a8825998b1dcf53] [ERROR] Errors: 
[f77ff50d2bcf5e2a9421364f8a8825998b1dcf53] [ERROR]   SetMessagesMethodTest.lambda$setMessagesShouldCreateMessageInOutboxWhenSendingMessage$3:2326 » ConcurrentModification
```

The party is not over

**Comment by mbaechler** [comment]
It failed into MemorySetMessagesMethodTest :
```
 [ERROR] Tests run: 130, Failures: 0, Errors: 1, Skipped: 1, Time elapsed: 201.832 s <<< FAILURE! - in org.apache.james.jmap.memory.MemorySetMessagesMethodTest
```

Maybe I'm wrong but it looks really unrelated with this PR.

test this please

**Comment by mbaechler** [comment]
SpamAssassin errors are back:

```
[f77ff50d2bcf5e2a9421364f8a8825998b1dcf53] [ERROR] Errors: 
[f77ff50d2bcf5e2a9421364f8a8825998b1dcf53] [ERROR]   SpamAssassinContract.deletingSpamMessageShouldNotImpactSpamAssassinState:557 » ConditionTimeout
[f77ff50d2bcf5e2a9421364f8a8825998b1dcf53] [ERROR]   SpamAssassinContract.expungingSpamMessageShouldNotImpactSpamAssassinState:493 » ConditionTimeout
[f77ff50d2bcf5e2a9421364f8a8825998b1dcf53] [ERROR]   SpamAssassinContract.imapCopiesToSpamMailboxShouldBeConsideredAsSpam:194 » ConditionTimeout
[f77ff50d2bcf5e2a9421364f8a8825998b1dcf53] [ERROR]   SpamAssassinContract.imapMovesToSpamMailboxShouldBeConsideredAsSpam:242 » ConditionTimeout
[f77ff50d2bcf5e2a9421364f8a8825998b1dcf53] [ERROR]   SpamAssassinContract.movingAMailToTrashShouldNotImpactSpamassassinLearning:370 » ConditionTimeout
[f77ff50d2bcf5e2a9421364f8a8825998b1dcf53] [ERROR]   SpamAssassinContract.spamShouldBeDeliveredInSpamMailboxOrInboxWhenMultipleRecipientsConfigurations:628 » ConditionTimeout
[f77ff50d2bcf5e2a9421364f8a8825998b1dcf53] [ERROR]   SpamAssassinContract.spamShouldBeDeliveredInSpamMailboxWhenSameMessageHasAlreadyBeenMovedToSpam:146 » ConditionTimeout
```

test this please

**Comment by mbaechler** [comment]
Ok, so it's green enough IMO, in 5 runs we had 2 unrelated failures.
Last step before merge is bench.

**Comment by mbaechler** [comment]
Before:

screenshot from 2019-02-01 18-38-55


After: 

screenshot from 2019-02-01 18-39-06

I think it's safe to merge (we are 4 times slower than on 14-12-18, we definitely need to fix the source of this slowdown)

**Comment by mbaechler** [comment]
merged, finally !

**Pull request closed without merging**